src/NITTA/Synthesis/Steps/Allocation.hs

```diff
@@ -0,0 +1,92 @@
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE NoMonomorphismRestriction #-}
+
+{-# OPTIONS -fno-warn-orphans #-}
+
+{- |
+Module      : NITTA.Synthesis.Steps.Allocation
+Description : Implementation of SynthesisDecisionCls that allows to allocate PUs
+Copyright   : (c) Aleksandr Penskoi, 2022
+License     : BSD3
+Maintainer  : aleksandr.penskoi@gmail.com
+Stability   : experimental
+-}
+module NITTA.Synthesis.Steps.Allocation (
+    AllocationMetrics (..),
+) where
+
+import Data.Aeson (ToJSON)
+import qualified Data.Map as M
+import GHC.Generics (Generic)
+import NITTA.Intermediate.Analysis (ProcessWave (ProcessWave, pwFs))
+import NITTA.Model.Networks.Bus (
+    BusNetwork (bnPUPrototypes, bnPus, bnRemains),
+    PUPrototype (PUPrototype, pProto),
+ )
+import NITTA.Model.Networks.Types (PU (PU, unit))
+import NITTA.Model.Problems.Allocation (
+    Allocation (Allocation, puTag),
+    AllocationProblem (allocationDecision),
+ )
+import NITTA.Model.ProcessorUnits.Types (
+    ParallelismType (..),
+    ProcessorUnit (parallelismType),
+    UnitTag,
+    allowToProcess,
+ )
+import NITTA.Model.TargetSystem (TargetSystem (TargetSystem, mUnit))
+import NITTA.Synthesis.Types (
+    SynthesisDecisionCls (..),
+    SynthesisState (SynthesisState, numberOfProcessWaves, processWaves, sTarget),
+ )
+
+data AllocationMetrics = AllocationMetrics
+    { -- |PU prototype parallelism type
+      mParallelism :: ParallelismType
+    , -- |The number of remaining functions that can be bound to pu
+      mRelatedRemains :: Int
```

Comment:
It is up to you, but maybe it will be more friendly for metrics if you use only Float type here. It allows you to avoid typecasting in the `estimate` function.